dotnet/Miscreant/Utils.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography;
+
+namespace Miscreant
+{
+	internal static class Utils
+	{
+		private static readonly byte[] deBruijn = new byte[] {
+			0, 1, 28, 2, 29, 14, 24, 3, 30, 22, 20, 15, 25, 17, 4, 8,
+			31, 27, 13, 23, 21, 19, 16, 7, 26, 12, 18, 6, 11, 5, 10, 9
+		};
+
+		private static readonly RandomNumberGenerator random = RandomNumberGenerator.Create();
+
+		public static ArraySegment<T> Slice<T>(this ArraySegment<T> seg, int index)
+		{
+			return new ArraySegment<T>(seg.Array, seg.Offset + index, seg.Count - index);
+		}
+
+		public static void Multiply(byte[] input)
+		{
+			Debug.Assert(input.Length == Constants.BlockSize);
+
+			int carry = input[0] >> 7;
+
+			for (int i = 0; i < Constants.BlockSize - 1; ++i)
+			{
+				input[i] = (byte)((input[i] << 1) | (input[i + 1] >> 7));
+			}
+
+			byte last = (byte)((input[Constants.BlockSize - 1] << 1) ^ ((0 - carry) & Constants.R));
+			input[Constants.BlockSize - 1] = last;
+		}
+
+		public static void Xor(byte[] source, byte[] destination, int length)
+		{
+			Xor(source, 0, destination, 0, length);
+		}
+
+		public static void Xor(byte[] source, int sourceIndex, byte[] destination, int destinationIndex, int length)
+		{
+			for (int i = 0; i < length; ++i)
+			{
+				destination[destinationIndex + i] ^= source[sourceIndex + i];
+			}
+		}
+
+		public static void Pad(byte[] buffer, int position)
+		{
+			buffer[position] = 0x80;
+
+			for (int i = position + 1; i < Constants.BlockSize; ++i)
+			{
+				buffer[i] = 0;
+			}
+		}
+
+		public static bool ConstantTimeEquals(byte[] x, byte[] y, int count)
```

Comment:
I have been told you should add

```
[MethodImpl(MethodImplOptions.NoOptimization | MethodImplOptions.NoInlining)]
```

...to methods purported to be constant time.

Reply:
I'll read the docs about the guarantees these flags provide and get back to you on this one tomorrow.

Reply:
`MethodImplOptions.NoOptimization` is definitively necessary. Not sure how `NoInlining` is useful in constant time scenarios, but Microsoft is using it in the standard library (see [this](https://github.com/aspnet/Identity/blob/dev/src/Microsoft.Extensions.Identity.Core/PasswordHasher.cs#L69) for example), so I added them both.